active/0000-stateful-helpers.md

```diff
@@ -0,0 +1,276 @@
+- RFC PR: (leave this empty)
+- Ember Issue: (leave this empty)
+
+# Summary
+
+Stateful helpers are a class-based way to define helpers. Stateful helpers:
+
+  * Have a single return value
+  * Can store and read state
+  * Have lifecycle hooks analagous to components where appropriate. For
+    example, a stateful helper may call `recompute` at any time to generate a new
+    value.
+  * Are a superset of traditional function-based helpers. They can do more,
+    but in many cases a traditional helper is appropriate.
+
+They fill a gap in Ember's current template APIs:
+
+ |has positional params|has template|has lifecycle, instance|can control rerender
+---|---|---|---|---
+components|Yes|Yes|Yes|Yes
+function helpers|Yes|No|No|No
+stateful helpers|Yes|No|Yes|Yes
+
+Example usage:
+
+```js
+// app/helpers/full-name.js
+import Ember from "ember";
+
+export default Ember.Helper.extend({
+  nameBuilder: Ember.inject.service(),
+  compute(params) {
+    const builder = this.get('nameBuilder');
+    return builder.fullName(params[0], params[1]);
+  }
+});
+```
+
+Use an stateful helper anywhere a traditional helper is valid:
+
+```hbs
+{{full-name 'Bigtime' 'Beagle'}}
+{{input value=(full-name 'Gyro' 'Gearloose') readonly=true}}
+```
+
+# Motivation
+
+Helpers in Ember are pure functions. This make them easy to reason about, but
+also overly simplistic. It is difficult to write a helper that recomputes due to
+something besides the change of its input, and helpers have no API for
+accessing parts of Ember like services.
+
+Specifically, this addresses many of the concerns in
+[emberjs/ember.js#11080](https://github.com/emberjs/ember.js/issues/11080).
+Libraries such as [yahoo/ember-intl](https://github.com/yahoo/ember-intl),
+[dockyard/ember-cli-i18n](https://github.com/dockyard/ember-cli-i18n), and
+[minutebase/ember-can](https://github.com/minutebase/ember-can) will be
+provided a viable public API to couple to.
+
+# Detailed design
+
+Stateful helpers have the same naming requirements, usage, and file placement
+as helpers today. They have similar lifecycle hooks to components.
+
+### Naming and file placement
+
+The requirement for naming a stateful helper are the same as those for existing
+helpers:
+
+* Helper names should use a `-`.
+* Helper names that do not use a `-` must be registered via
+  `Ember.Helpers.registerHelper`.
+* In Ember-CLI, helpers live in `app/helpers/full-name.js`, or with pods in
+  `app/full-name/helper.js`.
+
+### Defininition and lifecycle
+
+A stateful helper is defined as a class inheriting from `Ember.Helper`. For
+example:
+
+```js
+// app/helpers/hello-world.js
+import Ember from "ember";
+
+// Usage: {{hello-world}}
+export default Ember.Helper.extend({
+  compute() {
+    return "Hello Helper World";
+  }
+});
+```
+
+Upon initial render:
+
+* The helper instance is created.
+* The `compute` method is called, and its return value is treated like the
+  return value of a function-based helper.
```

Comment:
I've added a section on return values, and ran some tests. I discovered that function-based helpers don't work in some of the places I describe they should. I'd like to make them work in those cases.